src/ucp/rndv/proto_rndv.c

```diff
@@ -303,13 +335,161 @@ ucp_proto_rndv_bulk_init(const ucp_proto_multi_init_params_t *init_params)
     return UCS_OK;
 }
 
+size_t ucp_proto_rndv_pack_ack(void *dest, void *arg)
+{
+    ucp_request_t *req       = arg;
+    ucp_reply_hdr_t *ack_hdr = dest;
+
+    ack_hdr->req_id = req->send.rndv.remote_req_id;
+    ack_hdr->status = UCS_OK;
+
+    return sizeof(*ack_hdr);
+}
+
 void ucp_proto_rndv_bulk_config_str(size_t min_length, size_t max_length,
                                     const void *priv, ucs_string_buffer_t *strb)
 {
     const ucp_proto_rndv_bulk_priv_t *rpriv = priv;
 
     ucp_proto_multi_config_str(min_length, max_length, &rpriv->mpriv, strb);
-    ucs_string_buffer_appendf(strb, " am-ln:%d", rpriv->am_lane);
+    ucs_string_buffer_appendf(strb, " ");
+    ucp_proto_rndv_ack_config_str(min_length, max_length, &rpriv->super, strb);
+}
+
+static ucs_status_t
+ucp_proto_rndv_send_reply(ucp_worker_h worker, ucp_request_t *req,
+                          ucp_operation_id_t op_id, uint8_t sg_count,
+                          const void *rkey_buffer, size_t length)
+{
+    ucp_worker_cfg_index_t rkey_cfg_index;
+    ucp_proto_select_param_t sel_param;
+    ucp_proto_select_t *proto_select;
+    ucs_status_t status;
+    ucp_rkey_h rkey;
+
+    ucs_assert(op_id == UCP_OP_ID_RNDV_RECV);
+    ucs_assert(sg_count == 1);
+
+    if (rkey_buffer == NULL) {
+        /* No remote key, use endpoint protocols */
+        proto_select   = &ucp_ep_config(req->send.ep)->proto_select;
+        rkey_cfg_index = UCP_WORKER_CFG_INDEX_NULL;
+        rkey           = NULL;
+    } else {
+        status = ucp_ep_rkey_unpack(req->send.ep, rkey_buffer, &rkey);
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        proto_select   = &ucp_rkey_config(worker, rkey)->proto_select;
+        rkey_cfg_index = rkey->cfg_index;
+    }
+
+    ucp_proto_select_param_init(&sel_param, op_id, 0,
+                                req->send.state.dt_iter.dt_class,
+                                &req->send.state.dt_iter.mem_info, sg_count);
+
+    status = ucp_proto_request_set_proto(worker, req->send.ep, req,
+                                         proto_select, rkey_cfg_index,
+                                         &sel_param, length);
+    if (status != UCS_OK) {
+        goto err_destroy_rkey;
+    }
+
+    req->send.rndv.rkey = rkey;
+
+    ucp_trace_req(req,
+                  "%s rva 0x%" PRIx64 " rreq_id 0x%" PRIx64 " with protocol %s",
+                  ucp_operation_names[op_id], req->send.rndv.remote_address,
+                  req->send.rndv.remote_req_id,
+                  req->send.proto_config->proto->name);
+
+    ucp_request_send(req, 0);
+    return UCS_OK;
+
+err_destroy_rkey:
+    if (rkey != NULL) {
```

Comment:
seems rkey can be non NULL here if `ucp_ep_rkey_unpack` fails

Reply:
if ucp_ep_rkey_unpack fails we should jump to `err:` which is after this label

Reply:
right, my fault